core/src/main/scala/kafka/security/auth/Acl.scala

```diff
@@ -68,12 +75,28 @@ object Acl {
         val operation = Operation.fromString(itemJs(OperationKey).to[String])
         new Acl(principal, permissionType, host, operation)
       }.toSet
-    }.getOrElse(Set.empty)
+    }.right.getOrElse(Set.empty)
   }
 
   def toJsonCompatibleMap(acls: Set[Acl]): Map[String, Any] = {
     Map(Acl.VersionKey -> Acl.CurrentVersion, Acl.AclsKey -> acls.map(acl => acl.toMap.asJava).toList.asJava)
   }
+
+  /**
+    * Parse a JSON string into a JsonValue if possible. `None` is returned if `input` is not valid JSON. This method is currently used
+    * to read the already stored invalid ACLs JSON which was persisted using older versions of Kafka (prior to Kafka 1.1.0). KAFKA-6319
+    */
+  private def tryParseBytesIncludingACLs(input: Array[Byte]): Either[JsonProcessingException, JsonValue] =
+    try Right(mapper.readTree(input)).right.map(JsonValue(_))
+    catch {
+      case _: JsonProcessingException =>
+        // Before 1.0.1, Json#encode did not escape backslash or any other special characters. SSL principals
+        // stored in ACLs may contain backslash as an escape char, making the JSON generated in earlier versions invalid.
+        // Escape backslash and retry to handle these strings which may have been persisted in ZK.
+        // Note that this does not handle all special characters (e.g. non-escaped double quotes are not supported)
+        val escapedInput = new String(input, StandardCharsets.UTF_8).replaceAll("\\\\", "\\\\\\\\")
+        Json.tryParseBytes(escapedInput.getBytes(StandardCharsets.UTF_8))
+    }
```

Comment:
I think this can be written more simply as:

```scala
  private def parseBytesWithAclFallback(input: Array[Byte]): Option[JsonValue] = {
    Json.tryParseBytes(input) match {
      case Left(e) => //fallback code goes here
      case Right(v) => Some(v)
   }
```

Reply:
Thanks for the tip, I'll incorporate this in the PR.